### ISSUE TITLE
Memory Leak in LinkHandler?

### ISSUE THREAD

**Issue by GoogleCodeExporter** [mikaelgrev/miglayout#20]
```
We have an application where we are instantiating a lot of (~2000+) instances 
of MigLayout.

For each instance there are created at least one HashMap in LinkHandler. Those 
HashMaps are never cleaned up in our application.

Debugging showed me, that net.miginfocom.layout.LinkHandler#getValue() is never 
called.
I have added some code that calls the read method regularly, which fixes this 
issue. 


BTW:
net.miginfocom.layout.LinkHandler#clearWeakReferencesNow seems to not clear 
VALUES and VALUES_TEMP. 
I don't see any way those collections are cleaned up when 
clearWeakReferencesNow() has been called.

```

Original issue reported on code.google.com by `js.cedar...@gmail.com` on 28 Jan 2015 at 9:14

**Comment by canteroferron** [comment]
Hello,

I have the same error. My application has 690M in the LinkHandler class. My version is 4.2.

**Comment by JFormDesigner** [comment]
I can confirm the memory leak.

The following test program creates 10000 temporary MigLayout instances, which adds 10000 items to the 3 static arrays in LinkHandler, but never removes any (LinkHandler.getValue() is never called).

VisualVM shows me ~10000 WeakReference instances (in LinkHandler.LAYOUTS) and ~20000 HashMap in int[] instances (in LinkHandler.VALUES and VALUES_TEMP)
```java
public class LinkHandlerMemoryLeakTest
{
	public static void main(String[] args) {
		JPanel panel = new JPanel();
		for (int i = 0; i < 10000; i++) {
			MigLayout migLayout = new MigLayout();
			panel.setLayout(migLayout);
			panel.doLayout();
		}
		JOptionPane.showMessageDialog( null, "use VisualVM to check memory heap" );
	}
}
```

I'm working on a PR to fix this leak...

The idea is to replace the 3 static arrays in LinkHandler with a single WeakHashMap instance, where layout is used as key and the value is an array that contains two HashMaps (for VALUES and VALUES_TEMP). Then the GC can free all resources as soon as the MigLayout instance is no longer referenced.

**Comment by mikaelgrev** [comment]
Thank you Karl, that would be fantastic! :)

**Comment by JFormDesigner** [comment]
@canteroferron can you please test the [5.1-SNAPSHOT](https://oss.sonatype.org/content/repositories/snapshots/com/miglayout/) and report whether the problem is gone (so that we can close this issue).

**Comment by Paul-Pech** [comment]
Hi,

while testing our application for memory leaks, we too ran into the LinkHandler memory leak issue described here. We were still using Miglayout 4.3 and after reading this report, we switched to the 5.1 branch.

When we tested with 5.1 20160120-298 (core/swing) the leak seemed to be gone.

When we use 5.1 20160126-304 the leak seems to be back again (cf. attached screenshots of Eclipse MAT). An automated test ran for about 1 hour and created numerous nested JPanels each with a MigLayout. With v304 the LinkHandler occupies more than 5.5MB.

Running the same test with v298 results in substantially less memory used (less than 500k)

Thanks for MigLayout, a great Layout Manager indeed. It saved us a ton of work over the years!

Yours

Paul

![appclassloader](https://cloud.githubusercontent.com/assets/7048968/12676952/f5a1df72-c695-11e5-9ecc-96114d912573.png)
![0xf099efd8](https://cloud.githubusercontent.com/assets/7048968/12676955/fb5067e0-c695-11e5-9fa4-587eea336998.png)

**Comment by JFormDesigner** [comment]
Hi Paul,

can you please check where the JTables are referenced from?

The WeakHashMap$Entry (in the screenshot) is a subclass of WeakReference and should not block the GC from disposing the JTables. Maybe the problem is somewhere outside of LinkHandler.

BTW it is very strange that the WeakHashMap LinkHandler.LAYOUTS contains JTables.
It should always have MigLayout instances as key, but newer components. And the values are always HashMaps, which use Strings as key and int[] as values. So there should be no components anywhere in LinkHandler...

Karl

**Comment by Paul-Pech** [comment]
Hi Karl,

thank you for your fast answer.

I'm sorry to say that my bug report was incorrect. I can no longer reproduce the issue I described here. Before I opened the bug report, I ran the test for at least three times, always finding said memory leak.

Trying again today, I'm unable to reproduce any of it. Maybe a partial update from 4.3 to 5.1 is to blame, but most likely it was just me on a Friday afternoon...

Paul
